python/cuml/internals/input_utils.py

```diff
@@ -0,0 +1,516 @@
+
+
+def determine_array_type_full(X):
+    """
+    Returns a tuple of the array type, and a boolean if it is sparse
+
+    Parameters
+    ----------
+    X : array-like
+        Input array to test
+
+    Returns
+    -------
+    (string, bool) Returns a tuple of the array type string and a boolean if it
+        is a sparse array.
+    """
+    if (X is None):
+        return None, None
+
+    # Get the generic type
+    gen_type = get_supported_input_type(X)
+
+    if (gen_type is None):
+        return None, None
+
+    return _input_type_to_str[gen_type], gen_type in _sparse_types
+
+
+def is_array_like(X):
+    try:
+        return (
+            hasattr(X, '__cuda_array_interface__')
+            or (
+                hasattr(X, '__array_interface__')
+                and not (
+                    isinstance(X, global_settings.xpy.generic)
+                    or isinstance(X, type)
+                )
+            ) or isinstance(X, (
+                SparseCumlArray, CudfSeries, PandasSeries, CudfDataFrame,
+                PandasDataFrame))
+            or numba_cuda.devicearray.is_cuda_ndarray(X)
+        )
+    except UnavailableError:
+        return False
+
+
+@nvtx_annotate(message="common.input_utils.input_to_cuml_array",
+               category="utils", domain="cuml_python")
+def input_to_cuml_array(X,
+                        order='F',
+                        deepcopy=False,
+                        check_dtype=False,
+                        convert_to_dtype=False,
+                        check_mem_type=False,
+                        convert_to_mem_type=None,
+                        safe_dtype_conversion=True,
+                        check_cols=False,
+                        check_rows=False,
+                        fail_on_order=False,
+                        force_contiguous=True):
+    """
+    Convert input X to CumlArray.
+
+    Acceptable input formats:
+
+    * cuDF Dataframe - returns a deep copy always.
+    * cuDF Series - returns by reference or a deep copy depending on
+        `deepcopy`.
+    * Numpy array - returns a copy in device always
+    * cuda array interface compliant array (like Cupy) - returns a
+        reference unless `deepcopy`=True.
+    * numba device array - returns a reference unless deepcopy=True
+
+    Parameters
+    ----------
+
+    X : cuDF.DataFrame, cuDF.Series, NumPy array, Pandas DataFrame, Pandas
+        Series or any cuda_array_interface (CAI) compliant array like CuPy,
+        Numba or pytorch.
+
+    order: 'F', 'C' or 'K' (default: 'F')
+        Whether to return a F-major ('F'),  C-major ('C') array or Keep ('K')
+        the order of X. Used to check the order of the input. If
+        fail_on_order=True, the method will raise ValueError,
+        otherwise it will convert X to be of order `order` if needed.
+
+    deepcopy: boolean (default: False)
+        Set to True to always return a deep copy of X.
+
+    check_dtype: np.dtype (default: False)
```

Comment:
The next couple of comments are suggestions for API improvements which would likely fall outside of the scope of this PR.

```suggestion
    expect_dtype: np.dtype (default: None)
```

Reply:
Let's raise issues for this and anything else along these lines.